### PR TITLE
page_service: rewrite batching to work without a timeout

### PR DESCRIPTION
**Pull request by problame**
# Problem

The timeout-based batching adds latency to unbatchable workloads.

We can choose a short batching timeout (e.g. 10us) but that requires high-resolution timers, which tokio doesn't have.
I thoroughly explored options to use OS timers (see [this](https://github.com/neondatabase/neon/pull/9822) abandoned PR).
In short, it's not an attractive option because any timer implementation adds non-trivial overheads.

# Solution

The insight is that, in the steady state of a batchable workload, the time we spend in `get_vectored` will be hundreds of microseconds anyway.

If we prepare the next batch concurrently to `get_vectored`, we will have a sizeable batch ready once `get_vectored` of the current batch is done and do not need an explicit timeout.

This can be reasonably described as **pipelining of the protocol handler**.

# Implementation

We model the sub-protocol handler for pagestream requests (`handle_pagrequests`) as two futures that form a pipeline:

2. Batching: read requests from the connection and fill the current batch
3. Execution: `take` the current batch, execute it using `get_vectored`, and send the response.

The Reading and Batching stage are connected through a new type of channel called `spsc_fold`.

See the long comment in the `handle_pagerequests_pipelined`  for details.

# Changes

- Refactor `handle_pagerequests`
    - separate functions for
        - reading one protocol message; produces a `BatchedFeMessage` with just one page request in it
        - batching; tried to merge an incoming `BatchedFeMessage` into an existing `BatchedFeMessage`; returns `None` on success and returns back the incoming message in case merging isn't possible
        - execution of a batched message
    - unify the timeline handle acquisition & request span construction; it now happen in the function that reads the protocol message
- Implement serial and pipelined model
    - serial: what we had before any of the batching changes
      - read one protocol message
      - execute protocol messages
    - pipelined: the design described above
      - optionality for execution of the pipeline: either via concurrent futures vs tokio tasks
- Pageserver config
  - remove batching timeout field
  - add ability to configure pipelining mode
  - add ability to limit max batch size for pipelined configurations (required for the rollout, cf https://github.com/neondatabase/cloud/issues/20620 )
  - ability to configure execution mode
- Tests
  - remove `batch_timeout` parametrization
  - rename `test_getpage_merge_smoke` to `test_throughput`
    - add parametrization to test different max batch sizes and execution moes
  - rename `test_timer_precision` to `test_latency`
  - rename the test case file to `test_page_service_batching.py`
  - better descriptions of what the tests actually do

## On the holding The `TimelineHandle` in the pending batch

While batching, we hold the `TimelineHandle` in the pending batch.
Therefore, the timeline will not finish shutting down while we're batching.

This is not a problem in practice because the concurrently ongoing `get_vectored` call will fail quickly with an error indicating that the timeline is shutting down.
This results in the Execution stage returning a `QueryError::Shutdown`, which causes the pipeline / entire page service connection to shut down.
This drops all references to the `Arc<Mutex<Option<Box<BatchedFeMessage>>>>` object, thereby dropping the contained `TimelineHandle`s.

- => fixes https://github.com/neondatabase/neon/issues/9850

# Performance

Local run of the benchmarks, results in [this empty commit](https://github.com/neondatabase/neon/pull/9851/commits/1cf5b1463f69ba5066cbb0713912aec7bb5579ad) in the PR branch.

Key take-aways:
* `concurrent-futures` and `tasks` deliver identical `batching_factor`
  * tail latency impact unknown, cf https://github.com/neondatabase/neon/issues/9837
* `concurrent-futures` has higher throughput than `tasks` in all workloads (=lower `time` metric)
* In unbatchable workloads, `concurrent-futures` has 5% higher `CPU-per-throughput`  than that of `tasks`, and 15% higher than that of `serial`. 
* In batchable-32 workload, `concurrent-futures` has 8% lower `CPU-per-throughput` than that of `tasks` (comparison to tput of `serial` is irrelevant)
* in unbatchable workloads, mean and tail latencies of `concurrent-futures` is practically identical to `serial`, whereas `tasks` adds 20-30us of overhead

Overall, `concurrent-futures` seems like a slightly more attractive choice.

# Rollout

This change is disabled-by-default.

Rollout plan:
- https://github.com/neondatabase/cloud/issues/20620

# Refs

- epic: https://github.com/neondatabase/neon/issues/9376
- this sub-task: https://github.com/neondatabase/neon/issues/9377
- the abandoned attempt to improve batching timeout resolution: https://github.com/neondatabase/neon/pull/9820
- closes https://github.com/neondatabase/neon/issues/9850
- fixes https://github.com/neondatabase/neon/issues/9835

### CODE REVIEW

File: pageserver/src/page_service.rs
Comment (VladLazar) [inline pageserver/src/page_service.rs:1324]: I'm not sure I see this.

Let's say that `protocol_pipelining_mode=PageServiceProtocolPipeliningMode::Tasks` and the call to `pagesteam_handle_batched_message`.

`handle_pagerequests_pipelined` returns an error, but `read_message_task` and `batcher_task` are still running. They'll run until a message is received an realise that downstream is dead. For the `ConcurrentFutures` case, I think it works fine.

File: pageserver/src/page_service.rs
Comment (VladLazar) [inline pageserver/src/page_service.rs:1166]: nit: would `PagestreamEnd` be a better name for this?

File: pageserver/src/page_service.rs
Comment (VladLazar) [inline pageserver/src/page_service.rs:1199]: nit: this is never `None` outside the loop below, right? Seems that we use `Option` here just for the `take` and `replace` interface.

File: pageserver/src/page_service.rs
Comment (VladLazar) [inline pageserver/src/page_service.rs:1276]: nit: does this deserve a comment mentioning that the channel has "memory" (i.e. if the change happens while we are processing things below, it still gets detected on the next call)?

File: pageserver/src/page_service.rs
Comment (VladLazar) [inline pageserver/src/page_service.rs:1170]: `watch::channel` already wraps the value in a read/write lock. Can we skip this Mutex?

File: pageserver/src/page_service.rs
Comment (VladLazar) [inline pageserver/src/page_service.rs:1230]: Can we wrap `wait_notified` inside the `Outcome::CannotBatchNeedWaitForExecutor` variant?

File: pageserver/src/page_service.rs
Comment (VladLazar) [inline pageserver/src/page_service.rs:756]: nit: a comment explaining that the return value is carry-over would be nice.

File: pageserver/src/page_service.rs
Comment (VladLazar) [inline pageserver/src/page_service.rs:805]: Can we lift this into a function to make it a bit more readable?

File: libs/pageserver_api/src/config.rs
Comment (VladLazar) [inline libs/pageserver_api/src/config.rs:418]: Can we enable this in a few test unit/python to get some rudimentary continuous test coverage?

File: pageserver/src/page_service.rs
Comment (VladLazar) [reply pageserver/src/page_service.rs:1324]: > For the ConcurrentFutures case, I think it works fine.

Actually, I'm not sure about that. `join!` polls the futures from a lambda so a failure in one of the futures doesn't bail the function (I think). This is problematic for cases where we want to kill the pagestream connection on the server side (e.g. wrong shard).

**Review by VladLazar — COMMENTED** [review]
Went through all of it minus benchmark. Flushing comments.

Logic looks good to me, but have some questions about control flow.

File: libs/pageserver_api/src/config.rs
Comment (problame) [reply libs/pageserver_api/src/config.rs:419]: Hm, initially I was thinking to revert back to `None` before merging - it's `Some()` right now so I get CI coverage.

Let's keep this conversation open and decide later.

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs:1170]: The `watch::Receiver` doesn't allow mutable access, so, yes, this Mutex is necessary in the current constellation.

I asked in #rust for a better primitive: https://neondb.slack.com/archives/C0277TKAJCA/p1732277778048099

File: libs/pageserver_api/src/config.rs
Comment (problame) [reply libs/pageserver_api/src/config.rs:418]: > Can we enable this in a few test unit/python to get some rudimentary continuous test coverage?

I don't get the question. If the default is `Some` here, so, we have the entire test suite as coverage.

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs:1324]: Hmm yes, correct. Will address this tomorrow.

File: libs/pageserver_api/src/config.rs
Comment (VladLazar) [reply libs/pageserver_api/src/config.rs:418]: I assumed you won't merge it with pipelining enabled. If that's the case, I am suggesting to parametrize some existing python tests to use pipelining (we'd also have to tweak the compute config).

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs]: implemented in https://github.com/neondatabase/neon/pull/9851/commits/9bf2618c45f4d13400ab915a49b05160c18e7f2f and adopted in https://github.com/neondatabase/neon/pull/9851/commits/a23abb2cc0c88abf2dc615e37ab3ddc56d953c73

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs:757]: obsolete as of https://github.com/neondatabase/neon/pull/9851#discussion_r1859032178

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs]: Can you check the new abstracted-out sync primitive and see whether this comment still applies? https://github.com/neondatabase/neon/pull/9851#discussion_r1859032178

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs:801]: I kinda like having the merging criteria all in one place.

We can refactor as part of unit testing
- https://github.com/neondatabase/neon/issues/9834

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs]: Obsolete as of https://github.com/neondatabase/neon/pull/9851#discussion_r1859032178

File: pageserver/src/page_service.rs
Comment (problame) [inline pageserver/src/page_service.rs:291]: The `timeline_handles` may be None in this case.

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs]: Done in recent pushes.

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs:294]: No longer true

File: libs/utils/src/sync/spsc_fold.rs
Comment (problame) [reply libs/utils/src/sync/spsc_fold.rs:182]: Right, there's no functional difference, just the state tracking for ... self-documentation and possibly debugging.

File: libs/pageserver_api/src/config.rs
Comment (problame) [inline libs/pageserver_api/src/config.rs:450]: Will revert this back to `serial` before merging. See rollout plan https://github.com/neondatabase/cloud/issues/20620

**Review by VladLazar — APPROVED** [review]
I think it's fine to merge given the cautious release plan. Nice job getting this into shape. This is tricky stuff!

File: pageserver/src/page_service.rs
Comment (problame) [reply pageserver/src/page_service.rs]: Excellent catch, fixed in ce83418aa49ac363d0af02b5ee943c2bf9cf0207

**Pull request merged**
